From 464cbe5153503bfd479ce457f177baa3dc0365d5 Mon Sep 17 00:00:00 2001
From: Carol Bouchard <caboucha@cisco.com>
Date: Tue, 14 Feb 2017 12:11:13 -0500
Subject: [PATCH] ML2 Nexus driver fails on SessionCloseError

The ML2 Nexus ncclient(ssh) driver fails on connect call to Nexus 9K.
Failure seen in logs is as follows:
  ncclient.transport.errors.SessionCloseError: Unexpected session close
This error is seen with more recent images of the N9K when handling
yang data was introduced.  As a result, our driver must specify
that we plan to use ssh with XML data.  This is done by adding
'ssh_subsystem_name': 'xmlagent' to device_params. This change was
also verified against a Nexus3K which had a very old image.

Change-Id: I0ae3e3f51cd3bec4647fd24a9715ae42f46ee0a6
Closes-bug:  #1664293
---
 .../ml2/drivers/cisco/nexus/nexus_network_driver.py     | 17 ++++++++++-------
 1 file changed, 10 insertions(+), 7 deletions(-)

diff --git a/networking_cisco/plugins/ml2/drivers/cisco/nexus/nexus_network_driver.py b/networking_cisco/plugins/ml2/drivers/cisco/nexus/nexus_network_driver.py
index 4e619e7..6ac685f 100644
--- a/networking_cisco/plugins/ml2/drivers/cisco/nexus/nexus_network_driver.py
+++ b/networking_cisco/plugins/ml2/drivers/cisco/nexus/nexus_network_driver.py
@@ -277,13 +277,16 @@ class CiscoNexusDriver(object):
         hostkey_verify = cfg.CONF.ml2_cisco.host_key_checks
         try:
             # With new ncclient version, we can pass device_params...
-            man = self.ncclient.connect(host=nexus_host,
-                                        port=nexus_ssh_port,
-                                        username=nexus_user,
-                                        password=nexus_password,
-                                        hostkey_verify=hostkey_verify,
-                                        timeout=30,
-                                        device_params={"name": "nexus"})
+            man = self.ncclient.connect(
+                host=nexus_host,
+                port=nexus_ssh_port,
+                username=nexus_user,
+                password=nexus_password,
+                hostkey_verify=hostkey_verify,
+                timeout=30,
+                device_params={
+                    "name": "nexus",
+                    'ssh_subsystem_name': 'xmlagent'})
         except Exception as e:
             # Raise a Neutron exception. Include a description of
             # the original ncclient exception.
-- 
2.7.4

